👌 IMPROVE: Allow SqliteTempBackend to read/write archives

chrisjsewell opened the pull request:
The key reason I want this, is to allow temporary profiles to be exported, which I think is pretty nice!

```python
$ ipython                                                                                    
Python 3.8.13 | packaged by conda-forge | (default, Mar 25 2022, 06:05:47) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.31.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from aiida import load_profile, engine, orm, plugins
   ...: from aiida.storage.sqlite_temp import SqliteTempBackend
   ...: 
   ...: %load_ext aiida
   ...: 
   ...: profile = load_profile(
   ...:     SqliteTempBackend.create_profile(
   ...:         'myprofile',
   ...:         options={
   ...:             'warnings.development_version': False,
   ...:             'runner.poll.interval': 1
   ...:         },
   ...:     ),
   ...: )

In [2]: orm.SinglefileData("/Users/chrisjsewell/Documents/GitHub/aiida_core_develop/Dockerfile").store()
Out[2]: <SinglefileData: uuid: 1d2921fb-17bc-47e6-8db0-dbb98d5b6bd0 (pk: 1)>

In [3]: %verdi archive create --all --force test.aiida
Report: 
Archive Parameters
--------------------  ----------
Path                  test.aiida
Version               main_0001
Compression           6

Inclusion rules
----------------------------  -----
Computers/Nodes/Groups/Users  All
Computer Authinfos            False
Node Comments                 True
Node Logs                     True

Report: Validating Nodes                                                                                                                                                                                 
Report: Creating archive with:
-----  -
users  1
nodes  1
-----  -
Report: Finalizing archive creation...                                                                                                                                                                   
Report: Archive created successfully
Success: wrote the export archive file to test.aiida
```

sphuber reviewed (commented):
Can you explain why this is necessary? Why can't the archive not deal with UUID4 keys when exporting the repository content?

The reason I am asking is that in most cases using `sha256` to compute the file hash is probably going to be slower compared to generating a simple UUID. This is problematic since the sandbox repository is used anytime a node is created and files are written to it, which is often. We would need a very good reason to slow this operation down.

To get a sense of the difference in speed, I wrote a quick benchmarking script to run in the shell:
```python
import pathlib
import random
import tempfile
import uuid

from aiida.repository.backend.sandbox import SandboxRepositoryBackend

with tempfile.TemporaryDirectory() as tmpdir:

    dirpath = pathlib.Path(tmpdir)
    dirpath_content = dirpath / 'content'
    dirpath_content.mkdir()

    # Create N number of files with normally distributed file sizes
    N = 200
    mu = 5 * 1024 * 1024
    sigma = 1024 * 1024

    for i in range(N):
        size = random.normalvariate(mu, sigma)
        nbytes = int(size) if size > 1 else 1024  # Ensure the number of bytes is positive
        content = random.randbytes(nbytes)
        (dirpath_content / str(uuid.uuid4())).write_bytes(content)

    def create_sandbox_repo():
        dirpath_repo = dirpath / str(uuid.uuid4())
        dirpath_repo.mkdir()
        repository = SandboxRepositoryBackend(dirpath_repo)

        for filepath in dirpath_content.iterdir():
            key = repository.put_object_from_file(filepath)

    % timeit create_sandbox_repo()

```
It essentially generates a directory of `N` files with their file size normally distributed with mean `mu` and standard deviation `sigma`. I ran this for a repo of small files and larger files

Parameters (`N`, `mu`, `sigma`) | `main`           | PR               |
--------------------------------|------------------|------------------|
(`200`, `5 * 1024 ** 2`, `1024 ** 2`) | 673 ms ± 19.7 ms | 1.26 s ± 10.2 ms |
(`10000`, `1024`, `100`) | 489 ms ± 23.4 ms | 508 ms ± 8.81 ms |

Of course this is just a quick benchmark, but I think it shows what we would think by intuition: for small file sizes the difference is minimal, but for larger files size (of the order of MB, which is not that big), the penalty is a factor of 2. This will increase as the file size increases.

chrisjsewell commented:
> We would need a very good reason to slow this operation down.

@sphuber I've created a separate `SandboxShaRepositoryBackend` subclass, for use with `SqliteTempBackend`, so this will no longer affect general operation 👍 

> Can you explain why this is necessary? Why can't the archive not deal with UUID4 keys when exporting the repository content?

Because of here: https://github.com/aiidateam/aiida-core/blob/1c4df022752d3d58628d75643da5e43c35c1414c/aiida/tools/archive/create.py#L577-L581

as it mentions, it would be difficult to convert e.g. `uuid4 ->  sha256` during the export process, because you would have to also somehow change every node's `repository_metadata`

sphuber commented:
> as it mentions, it would be difficult to convert e.g. `uuid4 -> sha256` during the export process, because you would have to also somehow change every node's `repository_metadata`

So there is no technical limitation, it simply isn't implemented. There is no reason that we cannot update the nodes attributes when exporting them, is there? 

I think that it should actually be implemented because otherwise you are forcing all storage backends to use sha256 for the keys, or they cannot be exported. Especially with storage backends now being pluginnable, this seems like an unreasonable limitation. We might just want to document that when implementing one, choose sha256 whenever possible, so that exporting is cheap. But in some implementations, you might not have control over this.

chrisjsewell commented:
> There is no reason that we cannot update the nodes attributes when exporting them, is there?

In theory yes, but it's certainly not trivial; because now you would have to add 

1. a step in the process (before node export) to generate a dict of all `current -> sha256` file keys (and have this in memory, which could be large for large profiles)
2. a step during the node export to (recursively) regenerate the `repository_metadata` and replace all keys, before exporting, which yeh would be pretty slow

Not something I have time for at the moment I'm afraid 😅

sphuber commented:
Agree that it wouldn't be trivial, but certainly possible and at some point nice to have. I think we could simply first export all the database content of the nodes to the archive, and then iterate over the nodes to copy their repository content to the archive. This could be streamed so not everything needs to be read into memory and when going through the `Repository` interface, you get the new key hierarchy back which can then be inserted into the `repository_metadata` column of the respective node. This can be done in batches as to not have too many inserts, but also not keep to many `repository_metadata` hierarchies in memory. I will open an issue with this feature request and providing some context to this PR. If you could just add a small test that shows that you can export from a temp backend, I will accept this PR.

chrisjsewell commented:
@sphuber I also just added the requisite methods necessary to import an archive into a temporary profile, and added a test for the round-trip (profile1 -> archive -> profile2)

chrisjsewell commented:
@sphuber in [69d4ea9](https://github.com/aiidateam/aiida-core/pull/5658/commits/69d4ea923d92a2ed2aa46f6dea92568ec560743a) I fixed a few places where I found that the backend was not being properly set.

This is why it run fine in isolation, but failed when running after any tests that had already loaded an aiida profile

chrisjsewell commented:
@sphuber tests are now parsing, so this is good to go

sphuber commented on aiida/storage/sqlite_temp/backend.py:
```diff
@@ -42,7 +47,8 @@ def create_profile(
         options: dict | None = None,
-        debug: bool = False
+        debug: bool = False,
+        repo_path: str | Path | None = None,
```
Why is this optional? Shouldn't this be required?

chrisjsewell replied:
This will default to the `SandboxFolder` default, i.e. a temporary directory, I felt that was a suitable default no?

sphuber replied:
I see know, just wasn't very clear since that happens in `get_repository`. Fine to keep, but maybe a comment in the docstring of the `create_profile` might be useful.

sphuber commented on aiida/storage/sqlite_temp/backend.py:
```diff
@@ -120,12 +127,13 @@ def get_session(self) -> Session:
         if self._repo is None:
             # to-do this does not seem to be removing the folder on garbage collection?
-            self._repo = SandboxRepositoryBackend(filepath=get_config_option('storage.sandbox') or None)
+            repo_path = self.profile.storage_config.get('repo_path')
```
The above to-do should be incorrect? At least the `SandboxRepositoryBackend` implements the `__del__` method to delete the folder on disk.

chrisjsewell replied:
yeh, I recall observing some times when, on exiting an ipython session, this did not appear to "activate", but I can give it another try...

sphuber commented on aiida/storage/sqlite_temp/backend.py:
```diff
@@ -120,12 +127,13 @@ def get_session(self) -> Session:
         if self._repo is None:
             # to-do this does not seem to be removing the folder on garbage collection?
-            self._repo = SandboxRepositoryBackend(filepath=get_config_option('storage.sandbox') or None)
+            repo_path = self.profile.storage_config.get('repo_path')
+            self._repo = SandboxShaRepositoryBackend(filepath=Path(repo_path) if repo_path else None)
```
If `repo_path` can be unset and you can accept `None`, wouldn't it be better to do:
```suggestion
            repo_path = self.profile.storage_config.get('repo_path', None)
            self._repo = SandboxShaRepositoryBackend(filepath=Path(repo_path) if repo_path else None)
```
Otherwise this could except

sphuber commented on aiida/storage/sqlite_temp/backend.py:
```diff
@@ -195,11 +203,122 @@ def get_info(self, detailed: bool = False) -> dict:
-        raise NotImplementedError
+    @staticmethod
+    @functools.lru_cache(maxsize=18)
+    def _get_mapper_from_entity(entity_type: EntityTypes, with_pk: bool):
```
Where does this method, `bulk_insert` and `bulk_update` come from all of a sudden. Why do they need to be implemented here all of a sudden when just changing the repo backend? From a quick look I don't see any repo specific code in these implementations, right?

chrisjsewell replied:
these methods are currently only required by the `create_archive` and `import_archive` functions, providing a backend agnostic way to perform these operations (obviously more performant than the typical singular insert/update).
So a backend will run fine in "normal" operations without them, but will fail for archive creation/import.

Yes the code is specific to a backend using sqlalchemy to interface with a database. 
The code is basically copied from `PsqlDosBackend`, since obviously they are both using sqlalchemy (just sqlite instead of postgres)

sphuber replied:
Haven't looked into the code in detail, but would it then maybe not almost make more sense to have `SqliteTempBackend` subclass `PsqlDosBackend` and then override certain methods? Probably would just have to change a few methods just to change the repository backend and some other minor things.

sphuber commented on aiida/tools/archive/create.py:
```diff
@@ -259,7 +259,7 @@ def create_archive(
         EXPORT_LOGGER.report('Test Run: Stopping before archive creation')
         keys = set(
-            orm.Node.collection(backend).iter_repo_keys(
+            orm.nodes.node.NodeCollection(orm.Node, backend).iter_repo_keys(
```
Was this causing a bug? If I look at the `Collection` constructor, if the base entity type is not explicitly defined, it would take the `_entity_base_cls` which is defined for `Node`. So why was this not working? Maybe the change should have simply been:
```suggestion
            orm.nodes.node.NodeCollection(backend=backend).iter_repo_keys(
```
since `backend` was being misinterpreted for as `entity_class` since it was the first and only positional argument?

sphuber replied:
As a complete aside (and not for this PR), we probably should probably change `Collection.__init__` to:
```python
class Collection(abc.ABC, Generic[EntityType]):

    def __init__(self, backend: Optional['StorageBackend'] = None, entity_class: Type[EntityType]) -> None:
        """ Construct a new entity collection.

        :param backend: the backend instance to get the collection for, or use the default
        :param entity_class: the entity type e.g. User, Computer, etc
        """
        from aiida.orm.implementation import StorageBackend

        self._entity_type = entity_class or self._entity_base_cls()
        assert issubclass(self._entity_type, self._entity_base_cls())

        self._backend = backend or get_manager().get_profile_storage()
        type_check(backend, StorageBackend, allow_none=True)
```
that is to say to swap `backend `and `entity_class`. Usually the entity cls will be defined in the subclass through the `_entity_base_cls` classmethod and it is the backend that we want to specify. So it makes more sense to have that as the first positional argument.

chrisjsewell replied:
So `orm.nodes.node.NodeCollection(backend=backend)` will definitely not work at present, because `NodeCollection` does not implement its own `__init__` and `Collection.__init__` explicitly requires the `entity_class`.

Your change indeed makes sense, that the `entity_class` should not be needed to initialise 👍 

and yeh the whole `orm.Node.collection(backend)` API, I feel is a little "clunky",
because I don't like that it is having to "silently" load the global backend first (which for a temp backend may not have even been loaded).

sphuber replied:
I don't understand how any of this is working properly now. You are right that `entity_class` is currently required in the `Collection` constructor, but then how are things like
```
class SqliteTempBackend:
    @cached_property
    def authinfos(self):
        return orm.SqliteAuthInfoCollection(self)
```
correct? The `SqliteAuthInfoCollection` is instantiated passing the `SqliteTempBackend` instance as the first argument, i.e., for `entity_class`, but the `SqliteAuthInfoCollection` would expect a `BackendAuthInfo` as per its `entity_cls` implementation? It seems that this code is trying to define the `backend` by passing itself, but it should be misinterpreted for the `entity_class`? If that is true I would expect a whole load of bugs which is clearly not happening so what is going on here?

sphuber replied:
Ah never mind, that collection actually references a `BackendCollection` which actually only takes `backend` as an argument. Ok, that should really be the way the front-end `Collection` works as well. Anyway, we can ignore this discussion for this PR as it is not relevant. I will have a look at simplifying the `Collection` implementation

sphuber commented on aiida/tools/archive/create.py:
```diff
@@ -568,9 +568,10 @@ def _stream_repo_files(
-        orm.Node.collection(backend).iter_repo_keys(filters={'id': {
-            'in': list(node_ids)
-        }}, batch_size=batch_size)
+        orm.nodes.node.NodeCollection(orm.Node, backend
+                                      ).iter_repo_keys(filters={'id': {
+                                          'in': list(node_ids)
+                                      }}, batch_size=batch_size)
```
Same here?
```suggestion
        orm.nodes.node.NodeCollection(backend=backend
                                      ).iter_repo_keys(filters={'id': {
                                          'in': list(node_ids)
                                      }}, batch_size=batch_size)
```

chrisjsewell replied:
yeh again, this won't work until your proposed change to `Collection.__init__` is implemented

sphuber commented on aiida/tools/archive/imports.py:
```diff
@@ -732,7 +732,7 @@ def _transform(row):
                 # to-do this is probably not the most efficient way to do this
                 uuid, new_mtime, new_comment = row
-                cmt = orm.Comment.collection.get(uuid=uuid)
+                cmt = orm.comments.CommentCollection(orm.Comment, backend).get(uuid=uuid)
```
```suggestion
                cmt = orm.comments.CommentCollection(backend=backend).get(uuid=uuid)
```

sphuber reviewed (changes requested):
Cheers @chrisjsewell , feature would be great to have. Just some questions/comments.

chrisjsewell commented:
Thanks for the review @sphuber, I've addressed your comments, and I think your proposed commits could only be implemented, once the `Collection.__init__` code has been changed no?

sphuber commented:
> Thanks for the review @sphuber, I've addressed your comments, and I think your proposed commits could only be implemented, once the `Collection.__init__` code has been changed no?

We can ignore the `Collection.__init__` question for now yeah. I have resolved all conversations that are done, but there remain three open ones.

sphuber commented:
Thanks @chrisjsewell
